frontend/js/src/artist-similarity/src/Data.tsx

```diff
@@ -0,0 +1,150 @@
+import React, {useState, useEffect} from "react";
+import SimilarArtistsGraph from "./SimilarArtistsGraph";
+import Input from "./Input";
+import tinycolor from "tinycolor2";
+
+type ArtistType = {
+    artist_mbid: string;
+    name: string;
+    comment: string;
+    type: string;
+    gender: string;
+    score?: number;
+    reference_mbid?: string;
+}
+    
+type MarkupResponseType = {
+    data: string;
+    type: "markup";
+}
+
+type DatasetResponseType = {
+    columns: Array<string>;
+    data: Array<ArtistType>;
+    type: "dataset";
+}
+
+type ApiResponseType = Array<MarkupResponseType | DatasetResponseType>;
+
+type NodeType = {
+    id: string;
+    artist_mbid: string;
+    size: number;
+    color: string;
+    score: number;
+}
+
+type LinkType = {
+    source: string;
+    target: string;
+    distance: number;
+}
+
+type GraphDataType = {
+    nodes: Array<NodeType>;
+    links: Array<LinkType>;
+}
+
+const colorGenerator = ():  tinycolor.Instance=> {
+    var color = tinycolor("hsv(" + Math.random() * 360 + ", 100%, 90%)");
+    return color;
+}
+
+const Data = () => {
+    const ARTIST_MBID = "8f6bd1e4-fbe1-4f50-aa9b-94c450ec0f11";
+    const SIMILAR_ARTISTS_LIMIT_VALUE = 18;
+    const BASE_URL = "https://labs.api.listenbrainz.org/similar-artists/json?algorithm=session_based_days_7500_session_300_contribution_5_threshold_10_limit_100_filter_True_skip_30&artist_mbid=";
+    const LINK_DIST_MULTIPLIER = 200;
+    const MIN_LINK_DIST = 50;
+    const MAIN_NODE_SIZE = 150;
+    const SIMILAR_NODE_SIZE = 85;
+    const BACKGROUND_ALPHA = 0.2;
+    const COLOR_MIX_WEIGHT = 0.3;
+    
+    var color1 = colorGenerator();
+    var color2 = colorGenerator();
+
+    const [similarArtists, setSimilarArtists] = useState<Array<ArtistType>>();
+    const [mainArtist, setMainArtist] = useState<ArtistType>();
+    const [similarArtistsLimit, setSimilarArtistsLimit] = useState(SIMILAR_ARTISTS_LIMIT_VALUE);
+    const [colors, setColors] = useState([color1, color2]);
+    const [artistMBID, setArtistMBID] = useState(ARTIST_MBID);
+
+    var scoreList: Array<number> = [];
+    
+    const fetchData = async (ARTIST_MBID: string): Promise<void> => {
+        try {
+          const response = await fetch(BASE_URL + ARTIST_MBID);
+          const data = await response.json();
+          console.log(data);
+          processData(data);
+        }
+        catch (error){
+          //Error message goes here.
+          alert("Something went wrong while loading information, please try again");
+        }
+    }
+    
+    const processData = (dataResponse: ApiResponseType): void => {
+        let mainArtistResponse = dataResponse[1];
+        setMainArtist(mainArtistResponse.type === "dataset" ? mainArtistResponse.data[0] : undefined);
+        
+        let similarArtistsResponce = dataResponse[3];
+        setSimilarArtists(similarArtistsResponce.type === "dataset" ? similarArtistsResponce.data.slice(0, similarArtistsLimit) : undefined);
+
```

Comment:
Here I would take another approach which might be more readable without having to know the order or format of the API response quite as much. Effectively the same result, but the code is easier to read in my opinion:

```
// Get the datasets out of the API response
const artistsData = dataResponse.filter(response =>
response.type === "dataset");
if(artistsData.length){
  setMainArtist(artistsData[0]);
const similarArtistsResponse = artistsData[1];
  if(similarArtistsResponse?.data?.length){
    setSimilarArtists(similarArtistsResponse.data.slice(0, similarArtistsLimit);
  }
}
```
At each point where i set a value in the state, there is a previous check to ensure it's not undefined, so we don't need the tertiary operator.
What do you think?

Reply:
I am trying out your suggestion, the only problem I encounter is that for some reason. Typescript detects artistData as Dataset or Markup. So when I try to access its data field it shows that it might be string.
```
Argument of type 'string | ArtistType' is not assignable to parameter of type 'SetStateAction<ArtistType | undefined>'.
  Type 'string' is not assignable to type 'SetStateAction<ArtistType | undefined>'.ts(2345)
```

Reply:
Brilliant! That looks perfect.
And thanks, I learned something today !

Reply:
The article mentions using find() is his example, and that made me realize that would also be ideal for us since we only want the first item.